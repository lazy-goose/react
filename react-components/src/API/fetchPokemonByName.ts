import PokeAPI from 'pokeapi-typescript';

export async function fetchPokemonByName(name: string) {
  if (!name?.length) {
    throw new Error('Unknown name');
  }
  return await PokeAPI.Pokemon.resolve(name);
}
