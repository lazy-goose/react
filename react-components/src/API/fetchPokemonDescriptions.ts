import PokeAPI, { IPokemon } from 'pokeapi-typescript';

export async function fetchPokemonDescriptions(names: string[]) {
  const pokemonPromises = names.map((n) => PokeAPI.Pokemon.resolve(n));
  const settledPokemons = await Promise.allSettled(pokemonPromises);
  const fulfilledPokemons = settledPokemons.filter(
    (p) => p.status === 'fulfilled'
  ) as PromiseFulfilledResult<IPokemon>[];
  return fulfilledPokemons.map((p) => p.value);
}
