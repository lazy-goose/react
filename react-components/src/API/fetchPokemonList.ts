import PokeAPI, { INamedApiResource, IPokemon } from 'pokeapi-typescript';

export async function fetchPokemonList() {
  return (await PokeAPI.Pokemon.listAll()).results;
}

export type PokemonList = INamedApiResource<IPokemon>[];
