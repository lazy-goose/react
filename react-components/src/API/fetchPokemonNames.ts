import PokeAPI from 'pokeapi-typescript';

export async function fetchPokemonNames() {
  const resourceList = await PokeAPI.Pokemon.listAll();
  return resourceList.results.map((r) => r.name);
}
