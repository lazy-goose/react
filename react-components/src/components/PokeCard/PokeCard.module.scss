@use '../../utils/sass/mixins.scss';

.PokeCard {
  border: 1px solid transparent;
  &.Active {
    pointer-events: none;
    border-color: var(--link);
  }
  @include mixins.hover {
    border-color: var(--link);
  }
}

.LinkWrapper {
  padding: 1.5rem 1rem;
  display: block;
  text-decoration: none;
  > *:not(:first-child) {
    margin-block: 1rem 0;
  }
}

.ImageContainer {
  display: flex;
  justify-content: center;
  position: relative;
}

.Image {
  font-size: 1.75rem;
  display: grid;
  align-items: center;
  &::before {
    position: absolute;
    left: 0;
  }
}

.ImageFallbackBlock {
  display: grid;
  place-content: center;
  font-size: 3.25em;
  font-weight: bold;
}

.Name {
  font-size: 2em;
  font-weight: bold;
}

.Types {
  font-size: 1.5em;
}
