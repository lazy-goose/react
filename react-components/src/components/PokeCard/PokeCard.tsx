import { NavLink, useSearchParams } from 'react-router-dom';
import s from './PokeCard.module.scss';
import jcn from '../../utils/joinClassNames';

type PokeCardProps = {
  name: string;
  types: string[];
  imageUrl: string;
  imageAlt?: string;
  className?: string;
};

export default function PokeCard(props: PokeCardProps) {
  const { name, types, imageUrl, imageAlt, className } = props;
  const [searchParams] = useSearchParams();
  return (
    <div className={jcn(s.PokeCard, className)}>
      <NavLink
        to={{ pathname: `pokemon/${name}`, search: searchParams.toString() }}
        className={s.LinkWrapper}
      >
        <div className={s.ImageContainer}>
          {imageUrl ? (
            <img
              className={s.Image}
              src={imageUrl}
              alt={imageAlt || `${name} pokemon`}
            />
          ) : (
            <div className={s.ImageFallbackBlock}>?</div>
          )}
        </div>
        <h3 className={s.Name}>{name}</h3>
        <p className={s.Types}>Types: {types.join(', ')}</p>
      </NavLink>
    </div>
  );
}
