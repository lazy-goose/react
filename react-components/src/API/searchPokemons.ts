import PokeAPI, { IPokemon } from 'pokeapi-typescript';
import { PokemonList } from './fetchPokemonList';

const randomBetween = (min: number, max: number) =>
  Math.floor(Math.random() * (max - min + 1) + min);

export async function searchPokemons(
  value: string,
  list: PokemonList,
  page = 1,
  limit = 100
) {
  const FROM = Math.floor((page - 1) * limit);
  const TO = FROM + limit;

  const filtered = list
    .map((i) => i.name)
    .filter((n) => n.startsWith(value.toLowerCase().trimEnd()));

  const pageLimited = filtered.slice(FROM, TO);

  if (!pageLimited.length) {
    return new Promise<[[], 1]>((resolve) =>
      setTimeout(() => resolve([[], 1]), randomBetween(200, 500))
    );
  }

  const descriptionPromises = pageLimited.map((n) =>
    PokeAPI.Pokemon.resolve(n)
  );
  const descriptions = (await Promise.allSettled(descriptionPromises)).filter(
    (p) => p.status === 'fulfilled'
  ) as PromiseFulfilledResult<IPokemon>[];

  return [descriptions.map((d) => d.value), filtered.length] as const;
}
