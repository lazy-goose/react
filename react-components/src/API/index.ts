import { fetchPokemonNames } from './fetchPokemonNames';
import { searchPokemons } from './searchPokemons';
import { fetchPokemonByName } from './fetchPokemonByName';

export { fetchPokemonNames, searchPokemons, fetchPokemonByName };
