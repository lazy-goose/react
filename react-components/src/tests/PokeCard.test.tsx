import { vi, describe, test, expect, beforeEach, SpyInstance } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent, { UserEvent } from '@testing-library/user-event';
import { MemoryRouter } from 'react-router';
import PokeCard from '../components/PokeCard/PokeCard';
import pokemon from './data/pikachu.json';
import renderApp, { mockPokeAPI } from './utils/renderApp';
import * as API from '../API';

const pokemonData = {
  name: pokemon.name,
  imageUrl: pokemon.sprites.front_default,
  imageAlt: pokemon.name,
  types: pokemon.types.map(({ type: { name } }) => name),
} as const;

describe('Tests for the Card component', () => {
  const renderPokeCard = (pokemon = pokemonData) =>
    render(
      <MemoryRouter>
        <PokeCard key={pokemon.name} {...pokemonData} />
      </MemoryRouter>
    );

  let fetchPokemonByName: SpyInstance;
  let user: UserEvent;

  beforeEach(() => {
    user = userEvent.setup();
    mockPokeAPI();
    fetchPokemonByName = vi.spyOn(API, 'fetchPokemonByName');
  });

  test('Ensure that the card component renders the relevant card data', async () => {
    renderPokeCard();
    const Card = await screen.findByTestId('pokemon-card');
    expect(Card).toMatchSnapshot();
  });

  test('Validate that clicking on a card opens a detailed card component', async () => {
    renderApp();
    const Cards = await screen.findAllByTestId('pokemon-card-link');
    await user.click(Cards[0]);
    const Aside = await screen.findByTestId('aside');
    expect(Aside).toBeInTheDocument();
  });

  test('Check that clicking triggers an additional API call to fetch detailed information', async () => {
    renderApp();
    const Cards = await screen.findAllByTestId('pokemon-card-link');
    await user.click(Cards[1]);
    expect(fetchPokemonByName).toBeCalled();
    await screen.findByTestId('pokemon-details-description');
  });
});
