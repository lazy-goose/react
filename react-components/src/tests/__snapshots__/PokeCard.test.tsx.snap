// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for the Card component > Ensure that the card component renders the relevant card data 1`] = `
<div
  class="_PokeCard_bf7019"
  data-testid="pokemon-card"
>
  <a
    class="_LinkWrapper_bf7019"
    data-testid="pokemon-card-link"
    href="/pokemon/pikachu"
  >
    <div
      class="_ImageContainer_bf7019"
    >
      <img
        alt="pikachu"
        class="_Image_bf7019"
        height="192"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/25.png"
        width="192"
      />
    </div>
    <h3
      class="_Name_bf7019"
    >
      pikachu
    </h3>
    <p
      class="_Types_bf7019"
    >
      Types: 
      electric
    </p>
  </a>
</div>
`;
